bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/BookieShell.java

```diff
@@ -1848,6 +1909,288 @@ public int runCmd(CommandLine cmdLine) throws Exception {
+                return false;
+            }
+
+            /*
+             * if ensembleSize is equal to writeQuorumSize, then ofcourse all
+             * the entries of this segment are supposed to be stored in this
+             * bookie. If this is last segment of the ledger and if the ledger
+             * is not closed (this is a corner case), then we have to return
+             * true. For more info. Check BOOKKEEPER-237 and BOOKKEEPER-325.
+             */
+            if ((lastSegment && !isLedgerClosed) || (ensembleSize == writeQuorumSize)) {
+                return true;
+            }
+
+            /*
+             * the following check is required because ensembleSize can be
+             * greater than writeQuorumSize and in this case if there are only
+             * couple of entries then based on RoundRobinDistributionSchedule
+             * there might not be any entry copy in this bookie though this
+             * bookie is part of the ensemble of this segment. If no entry is
+             * stored in this bookie then we should return false, because
+             * ReplicationWorker wont take care of fixing the ledgerMetadata of
+             * this segment in this case.
+             * 
+             * if ensembleSize > writeQuorumSize, then in LedgerFragment.java
+             * firstEntryID may not be equal to firstStoredEntryId lastEntryId
+             * may not be equalto lastStoredEntryId. firstStoredEntryId and
+             * lastStoredEntryId will be LedgerHandle.INVALID_ENTRY_ID, if no
+             * entry of this segment stored in this bookie. In this case
+             * LedgerChecker.verifyLedgerFragment will not consider it as
+             * unavailable/bad fragment though this bookie is part of the
+             * ensemble of the segment and it is down.
+             */
+            DistributionSchedule distributionSchedule = new RoundRobinDistributionSchedule(
+                    ledgerMetadata.getWriteQuorumSize(), ledgerMetadata.getAckQuorumSize(),
+                    ledgerMetadata.getEnsembleSize());
+            ArrayList<BookieSocketAddress> currentSegmentEnsemble = segments.get(segmentNo).getValue();
+            int thisBookieIndexInCurrentEnsemble = currentSegmentEnsemble.indexOf(bookieAddress);
+            long firstEntryId = segments.get(segmentNo).getKey();
+            long lastEntryId = lastSegment ? ledgerMetadata.getLastEntryId() : segments.get(segmentNo + 1).getKey() - 1;
+            long firstStoredEntryId = LedgerHandle.INVALID_ENTRY_ID;
+            long firstEntryIter = firstEntryId;
+            // following the same approach followed in LedgerFragment.getFirstStoredEntryId()
+            for (int i = 0; i < ensembleSize && firstEntryIter <= lastEntryId; i++) {
+                if (distributionSchedule.hasEntry(lastEntryId, thisBookieIndexInCurrentEnsemble)) {
```

Comment:
this has to be, 

hasEntry(firstEntryIter, thisBookieIndexInCurrentEnsemble) 

not

hasEntry(lastEntryId, thisBookieIndexInCurrentEnsemble)